api/API_VERSIONING.md

```diff
@@ -0,0 +1,206 @@
+general it requires new code and tests to be implemented in Envoy by hand. For example, if a developer
+wants to unify `HeaderMatcher` with `StringMatcher` in the route configuration, this is a likely
+candidate for this class of change. The following steps are required:
+1. The new version of the feature, e.g. the `NewHeaderMatcher` message should be added, together
+   with referencing fields, in the current stable major version for the route configuration proto.
+2. The Envoy implementation should be changed to consume configuration from the fields added in (1).
+   Translation code (and tests) should be written to map from the existing field and messages to
+   (1).
+3. The old message/enum/field/enum value should be annotated as deprecated.
+4. At the next major version, `protoxform` will remove the deprecated version automatically.
+
+This approach ensures that API major version releases are predictable and mechanical, and has the
+bulk of the Envoy code and test changes owned by feature developers, rather than the API owners.
+There will be no major `vN` initiative to address technical debt beyond that enabled by the above
+process.
+
+# One Definition Rule (ODR)
+
+To avoid maintaining more than two stable major versions of a package, and to cope with diamond
+dependency, we add a restriction on how packages may be referenced transitively; a package may have
+at most one version of another package in its transitive dependency set. This implies that some
+packages will have a major version bump during a release cycle simply to allow them to catch up to
+the current stable version of their dependencies.
+
+Some of this complexity and churn can be avoided by having strict rules on how packages may
+reference each other. Package organization and `BUILD` visibility constraints should be used
+restrictions to maintain a shallow depth in the dependency tree for any given package.
+
+# Minimizing the impact of churn
+
+In addition to stability, the API versioning policy has an explicit goal of minimizing the developer
+overhead for the Envoy community, other clients of the APIs (e.g. gRPC), management server vendors
+and the wider API tooling ecosystem. A certain amount of API churn between major versions is
+desirable to reduce technical debt and to support API evolution, but too much creates costs and
+barriers to upgrade.
+
+We consider deprecations to be *mandatory changes*. Any deprecation will be removed at the next
+stable API version.
+
+Other mechanical breaking changes are considered *discretionary*. These include changes such as
+field renames and are largely reflected in protobuf comments. The `protoxform` tool may decide to
+minimize API churn by deferring application of discretionary changes until a major version cycle
+where the respective message is undergoing a mandatory change.
+
+The Envoy API structure helps with minimizing churn between versions. Developers should architect
```

Comment:
@alyssawilk I think this covers what we discussed the other day, an advisory statement on organizing to minimize churn.